hr_attendance_management/models/hr_employee.py

```diff
@@ -56,31 +74,84 @@ def _compute_work_location(self):
-                employee.attendance_days_ids.mapped('extra_hours_lost') or [0])
+    def is_annual_at_date(self, date):
+        config = self.env['base.config.settings'].create({})
+        if config.get_penultimate_balance_cron_execution() \
+                < date \
+                < config.get_last_balance_cron_execution:
+            return self.previous_period_extra_hours_status
+        elif config.get_last_balance_cron_execution < date:
+            return self.extra_hours_status
+        else:
+            # undefined as we don't know the period and the status
+            raise ValueError("Date should be after the penultimate cron "
+                             "execution")
+
+    @api.multi
+    def complete_balance_computation(self, start_date=None, end_date=None,
+                                     existing_balance=0):
+        self.ensure_one()
+        config = self.env['base.config.settings'].create({})
+        max_extra_hours = self.env['base.config.settings'].create({})\
+            .get_max_extra_hours()
+        if not start_date:
+            start_date = fields.Date.to_string(
+                datetime.date.today().replace(month=1, day=1))
+        if not end_date:
+            end_date = fields.Date.to_string(datetime.date.today())
+
+        if start_date > end_date:
+            raise ValidationError("Start date must be earlier than end date.")
+        if start_date < config.get_penultimate_balance_cron_execution():
+            # too much in the past, we have no knowledge of annual status ->
+            # might yield incorrect results
+            _logger.info("Balance computation on outdated period which can "
+                         "result in incorrect results")
+
+        attendance_day_ids = self.attendance_days_ids.filtered(
+            lambda r: start_date <= r.date < end_date)
+
+        days = [start_date]
+        extra_hours_sum = [existing_balance]
+        lost_hours = [0]
+        attendance_days_sorted = sorted(attendance_day_ids,
+                                        key=lambda r: r.date)
+        for day in attendance_days_sorted:
+            days.append(day.date)
+            extra_hours_sum.append(extra_hours_sum[-1] + day.day_balance)
+            lost_hours.append(lost_hours[-1])
+            if extra_hours_sum[-1] > max_extra_hours and \
+                    not self.is_annual_at_date(day.date):
+                lost_hours[-1] += extra_hours_sum[-1] - max_extra_hours
+                extra_hours_sum[-1] = max_extra_hours
+
+        return days, extra_hours_sum, lost_hours
+
+    @api.multi
+    def past_balance_computation(self, start_date=None, end_date=None,
+                                 existing_balance=0):
+        _, extra_hours_sum, lost_hours = self.complete_balance_computation(
+            start_date=start_date,
+            end_date=end_date,
+            existing_balance=existing_balance)
+        return extra_hours_sum[-1], lost_hours[-1]
```

Comment:
The list returned will always contain at least one element? Otherwise this could fail.